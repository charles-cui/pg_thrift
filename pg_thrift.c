#include <postgres.h>
#include <port.h>
#include <catalog/pg_type.h>
#include <utils/builtins.h>
#include <utils/array.h>
#include <utils/lsyscache.h>

PG_MODULE_MAGIC;

PG_FUNCTION_INFO_V1(thrift_decode);
Datum thrift_decode(PG_FUNCTION_ARGS) {
  char* result_buff = "hello world";
  PG_RETURN_CSTRING((Datum)result_buff);
}
