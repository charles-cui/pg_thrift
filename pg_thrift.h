#ifndef _PG_THRIFT_H_
#define _PG_THRIFT_H_

#include <postgres.h>
#include <port.h>


#define THRIFT_RESULT_MAX_FIELDS 256

#define PG_THRIFT_TYPE_BOOL 2
#define PG_THRIFT_TYPE_BYTE 3
#define PG_THRIFT_TYPE_DOUBLE 4
#define PG_THRIFT_TYPE_INT16 6
#define PG_THRIFT_TYPE_INT32 8
#define PG_THRIFT_TYPE_INT64 10
#define PG_THRIFT_TYPE_STRING 11
#define PG_THRIFT_TYPE_STRUCT 12
#define PG_THRIFT_TYPE_MAP 13
#define PG_THRIFT_TYPE_SET 14
#define PG_THRIFT_TYPE_LIST 15

#define BYTE_LEN 4
#define PG_THRIFT_TYPE_LEN 1
#define PG_THRIFT_FIELD_LEN 2
#define DOUBLE_LEN 8
#define INT16_LEN 2
#define INT32_LEN 4
#define INT64_LEN 8
#define LIST_LEN 4
#define BOOL_LEN 1
#define FIELD_LEN 2
#define MAX_JSON_STRING_SIZE 1024

#endif // _PG_THRIFT_H_
