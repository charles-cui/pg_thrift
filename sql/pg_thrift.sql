CREATE EXTENSION pg_thrift;

-- struct (id = 123, phones=["123456", "abcdef"])
SELECT thrift_binary_get_int32(E'\\x0800010000007b0f00020b00000002000000063132333435360000000661626364656600' :: bytea, 1);

SELECT thrift_binary_get_bool(E'\\x0800010000007b0f00020b00000002000000063132333435360000000661626364656600' :: bytea, 1);

SELECT thrift_binary_get_byte(E'\\x0300010000000ce4bda0e5a5bde4b896e7958c00' :: bytea, 1);

-- hello world
SELECT thrift_binary_get_string(E'\\x0b00010000000ce4bda0e5a5bde4b896e7958c00' :: bytea, 1);

-- 1234567890.1234567890
SELECT thrift_binary_get_double(E'\\x04000141d26580b487e6b700' :: bytea, 1);

-- struct (id = 123, phones=["123456", "abcdef"])
SELECT thrift_binary_get_list_bytea(E'\\x0800010000007b0f00020b00000002000000063132333435360000000661626364656600' :: bytea, 2);

SELECT parse_thrift_binary_string(UNNEST(thrift_binary_get_list_bytea(E'\\x0800010000007b0f00020b00000002000000063132333435360000000661626364656600' :: bytea, 2)));

-- struct (id = true)
SELECT thrift_compact_get_bool(E'\\x1100' :: bytea, 1);

-- struct (id = false)
SELECT thrift_compact_get_bool(E'\\x1200' :: bytea, 1);

-- struct (id = 123, phones=["123456", "abcdef"])
SELECT thrift_compact_get_int16(E'\\x14f6011928063132333435360661626364656600' :: bytea, 1);

-- struct (id = 123, phones=["123456", "abcdef"])
SELECT thrift_compact_get_int32(E'\\x15f6011928063132333435360661626364656600' :: bytea, 1);

-- struct (id = 123, phones=["123456", "abcdef"])
SELECT thrift_compact_get_int64(E'\\x16f6011928063132333435360661626364656600' :: bytea, 1);

-- 你好世界
SELECT thrift_compact_get_byte(E'\\x1318e4bda0e5a5bde4b896e7958c00' :: bytea, 1);

-- 你好世界
SELECT thrift_compact_get_string(E'\\x1818e4bda0e5a5bde4b896e7958c00' :: bytea, 1);

-- 1234567890.1234567890
SELECT thrift_compact_get_double(E'\\x1741d26580b487e6b700' :: bytea, 1);

-- [1, 2, 3, 4, 5]
SELECT parse_thrift_compact_int16(UNNEST(thrift_compact_get_list_bytea(E'\\x1956020406080a00' :: bytea, 1)));

-- [1, 2, 3, 4, 5]
SELECT parse_thrift_compact_int32(UNNEST(thrift_compact_get_set_bytea(E'\\x1a58020406080a00' :: bytea, 1)));

DROP EXTENSION pg_thrift;
