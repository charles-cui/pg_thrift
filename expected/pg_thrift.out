CREATE EXTENSION pg_thrift;
-- struct (id = 123, phones=["123456", "abcdef"])
SELECT thrift_binary_get_int32(E'\\x0800010000007b0f00020b00000002000000063132333435360000000661626364656600' :: bytea, 1);
 thrift_binary_get_int32 
-------------------------
                     123
(1 row)

SELECT thrift_binary_get_bool(E'\\x0800010000007b0f00020b00000002000000063132333435360000000661626364656600' :: bytea, 1);
ERROR:  Invalid thrift format
SELECT thrift_binary_get_byte(E'\\x0300010000000ce4bda0e5a5bde4b896e7958c00' :: bytea, 1);
   thrift_binary_get_byte   
----------------------------
 \xe4bda0e5a5bde4b896e7958c
(1 row)

-- hello world
SELECT thrift_binary_get_string(E'\\x0b00010000000ce4bda0e5a5bde4b896e7958c00' :: bytea, 1);
 thrift_binary_get_string 
--------------------------
 你好世界
(1 row)

-- 1234567890.1234567890
SELECT thrift_binary_get_double(E'\\x04000141d26580b487e6b700' :: bytea, 1);
 thrift_binary_get_double 
--------------------------
         1234567890.12346
(1 row)

-- struct (id = 123, phones=["123456", "abcdef"])
SELECT thrift_binary_get_list_bytea(E'\\x0800010000007b0f00020b00000002000000063132333435360000000661626364656600' :: bytea, 2);
             thrift_binary_get_list_bytea              
-------------------------------------------------------
 {"\\x00000006313233343536","\\x00000006616263646566"}
(1 row)

SELECT parse_thrift_binary_string(UNNEST(thrift_binary_get_list_bytea(E'\\x0800010000007b0f00020b00000002000000063132333435360000000661626364656600' :: bytea, 2)));
 parse_thrift_binary_string 
----------------------------
 123456
 abcdef
(2 rows)

-- struct (id = true)
SELECT thrift_compact_get_bool(E'\\x1100' :: bytea, 1);
 thrift_compact_get_bool 
-------------------------
 t
(1 row)

-- struct (id = false)
SELECT thrift_compact_get_bool(E'\\x1200' :: bytea, 1);
 thrift_compact_get_bool 
-------------------------
 f
(1 row)

-- struct (id = 123, phones=["123456", "abcdef"])
SELECT thrift_compact_get_int16(E'\\x14f6011928063132333435360661626364656600' :: bytea, 1);
 thrift_compact_get_int16 
--------------------------
                      123
(1 row)

-- struct (id = 123, phones=["123456", "abcdef"])
SELECT thrift_compact_get_int32(E'\\x15f6011928063132333435360661626364656600' :: bytea, 1);
 thrift_compact_get_int32 
--------------------------
                      123
(1 row)

-- struct (id = 123, phones=["123456", "abcdef"])
SELECT thrift_compact_get_int64(E'\\x16f6011928063132333435360661626364656600' :: bytea, 1);
 thrift_compact_get_int64 
--------------------------
                      123
(1 row)

-- 你好世界
SELECT thrift_compact_get_byte(E'\\x1318e4bda0e5a5bde4b896e7958c00' :: bytea, 1);
  thrift_compact_get_byte   
----------------------------
 \xe4bda0e5a5bde4b896e7958c
(1 row)

-- 你好世界
SELECT thrift_compact_get_string(E'\\x1818e4bda0e5a5bde4b896e7958c00' :: bytea, 1);
 thrift_compact_get_string 
---------------------------
 你好世界
(1 row)

-- 1234567890.1234567890
SELECT thrift_compact_get_double(E'\\x1741d26580b487e6b700' :: bytea, 1);
 thrift_compact_get_double 
---------------------------
          1234567890.12346
(1 row)

-- [1, 2, 3, 4, 5]
SELECT parse_thrift_compact_int16(UNNEST(thrift_compact_get_list_bytea(E'\\x1956020406080a00' :: bytea, 1)));
 parse_thrift_compact_int16 
----------------------------
                          1
                          2
                          3
                          4
                          5
(5 rows)

-- [1, 2, 3, 4, 5]
SELECT parse_thrift_compact_int32(UNNEST(thrift_compact_get_set_bytea(E'\\x1a58020406080a00' :: bytea, 1)));
 parse_thrift_compact_int32 
----------------------------
                          1
                          2
                          3
                          4
                          5
(5 rows)

-- {'a' : 2}
SELECT parse_thrift_compact_string((thrift_compact_get_map_bytea(E'\\x1b02b602610400' :: bytea, 1))[1]);
 parse_thrift_compact_string 
-----------------------------
 a
(1 row)

-- {'a' : 2}
SELECT parse_thrift_compact_int16((thrift_compact_get_map_bytea(E'\\x1b02b602610400' :: bytea, 1))[2]);
 parse_thrift_compact_int16 
----------------------------
                          2
(1 row)

-- struct(id=123, phones=["123456", "abcdef"])
SELECT parse_thrift_compact_string(UNNEST(thrift_compact_get_list_bytea(E'\\x15f601192b0c3132333435360c61626364656600', 2)));
 parse_thrift_compact_string 
-----------------------------
 123456
 abcdef
(2 rows)

-- struct(id=123, phones=["123456", "abcdef"])
-- struct(id=456, phones=["123456", "abcdef"])
-- struct(id=123, items=[item1, item2])
SELECT parse_thrift_compact_string(UNNEST(thrift_compact_get_list_bytea(UNNEST(thrift_compact_get_list_bytea(E'\\x15f601192c15f601192b0c3132333435360c61626364656600159007192b0c3132333435360c6162636465660000', 2)), 2)));
 parse_thrift_compact_string 
-----------------------------
 123456
 abcdef
 123456
 abcdef
(4 rows)

DROP EXTENSION pg_thrift;
